Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/SNNClusterer.java

```diff
@@ -0,0 +1,637 @@
+import gov.sandia.cognition.math.matrix.Vectorizable;
+import gov.sandia.cognition.util.ObjectUtil;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+import java.util.Comparator;
+import java.util.RandomAccess;
+
+/**
+ * The Shared Nearest Neighbor algorithm is a variation of DBSCAN that clusters
+ * points with a high number of mutual near neighbors. SNN requires a distance
+ * metric, a cluster creator, an integer k (representing the number of nearest
+ * neighbors to use), an epsilon value (signifying the minimum number of shared
+ * neighbors two points need to be considered strongly connected), and a min
+ * points integer (serving as a minimum threshold of strongly connected
+ * neighbors for a point to be considered a core point). The algorithm uses the
+ * distance metric to calculate nearest neighbors, but then clusters based on
+ * mutual neighbors. This is done iteratively across all points, with any points
+ * that do not have sufficient strongly connected neighbors nor neighbors that
+ * are core points becoming noise points.
+ *
+ * @author mbain
+ * @param <DataType> The type of the data to cluster.
+ * @param <ClusterType> The type of {@code Cluster} created by the algorithm.
+ */
+@PublicationReference(
+    author =
+    {
+        "Levent Ertoz",
+        "Michael Steinbach",
+        "Vipin Kumar"
+    },
+    title = "Finding Clusters of Different Sizes, Shapes and Densities in "
+    + "Noisy, High Dimensional Data.",
+    type = PublicationType.Conference,
+    publication
+    = "Proceedings of Second SIAM Interantional Conference on Data Mining",
+    year = 2003,
+    url = "http://www-users.cs.umn.edu/~kumar/papers/SIAM_snn.pdf"
+)
+public class SNNClusterer<DataType extends Vectorizable, ClusterType extends Cluster<DataType>>
+    extends AbstractAnytimeBatchLearner<Collection<? extends DataType>, Collection<ClusterType>>
+    implements BatchClusterer<DataType, ClusterType>
+{
+
+    /**
+     * The default k is {@value}.
+     */
+    public static final int DEFAULT_K = 5;
+
+    /**
+     * The default eps is {@value}.
+     */
+    public static final int DEFAULT_EPS = 2;
+
+    /**
+     * The default min points is {@value}.
+     */
+    public static final int DEFAULT_MINPTS = 3;
+
+    /**
+     * The default maximum number of iterations {@value}
+     */
+    public static final int DEFAULT_MAX_ITERATIONS = Integer.MAX_VALUE;
+
+    /**
+     * The number of nearest neighbors.
+     */
+    private int k;
+
+    /**
+     * The similarity (shared nearest neighbors) score.
+     */
+    private int eps;
+
+    /**
+     * The minimum number of neighbors with a similarity score greater than or
+     * equal to eps to be considered a core point.
+     */
+    private int minPts;

```

Comment:
Minor naming: Can you spell out the name a bit more to make it clearer? minPoints? Or minNeighbors? (Here and elsewhere this is used)